lib/src/dbscan_base.dart

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2017, erikrahtjen. All rights reserved. Use of this source code
+// is governed by a BSD-style license that can be found in the LICENSE file.
+
+class DBSCAN<T> {
+  final List<T> _dataset;
+  final num _epsilon;
+  final num _minPts;
+  final Function _distance;
+
+  List<List<int>> _clusters = [];
+  List<int> _noise = [];
+  List<int> _visited = [];
+  List<int> _assigned = [];
+
+  DBSCAN(this._dataset, this._epsilon, this._minPts, this._distance(T a, T b)) {
+    _visited = new List(_dataset.length);
+    _assigned = new List(_dataset.length);
+
+    for (int pointId = 0; pointId < _dataset.length; pointId++) {
+      // if point is not visited, check if it forms a cluster
+      if (_visited[pointId] != 1) {
+        this._visited[pointId] = 1;
+
+        // if closest neighborhood is too small to form a cluster, mark as noise
+        var neighbors = _regionQuery(pointId);
+
+        if (neighbors.length < _minPts) {
+          _noise.add(pointId);
+        } else {
+          // create new cluster and add point
+          var clusterId = _clusters.length;
+          _clusters.add(new List());
```

Comment:
Does it make sense to introduce a Cluster type so you don't have to manage the list directly?